File: kotlin/day05.kt
package day05
import common.readInputLines

typealias Point = Pair<Int, Int>
typealias Line = Pair<Point, Point>

fun parsePt(pt: String): Point {
    val xy = pt.split(",").map {it.toInt()}
    return Pair(xy[0], xy[1])
}

fun parseLine(line: String): Line {
    val ab = line.split(" -> ").map(::parsePt)
    return Pair(ab[0], ab[1])
}

fun sgn(x: Int): Int {
    if (x == 0) return 0
    if (x < 0) return -1 else return 1
}

fun stroke(a: Point, b: Point, ptmap: HashMap<Point, Int>, skipDiagonals: Boolean) {
    val (x1, y1) = a
    val (x2, y2) = b
    val dx = sgn(x2 - x1)
    val dy = sgn(y2 - y1)
    if (skipDiagonals && dx != 0 && dy != 0) {
        return;
    }
    var cx = x1
    var cy = y1
    while (true) {
        val cpt = Pair(cx, cy)
        ptmap[cpt] = ptmap.getOrDefault(cpt, 0) + 1
        if (cx == x2 && cy == y2) {
            break;
        }
        cx += dx
        cy += dy
    }
}

fun findNumOverlaps(lines: List<Line>, skipDiagonals: Boolean): Int {
    var ptmap = HashMap<Point, Int>()
    for ((a, b) in lines) {
        stroke(a, b, ptmap, skipDiagonals);
    }
    return ptmap.map {(_, cnt) -> if (cnt >= 2) 1 else 0}.sum()
}

fun solution() {
    val lines = readInputLines(5).map(::parseLine)
    val res1 = findNumOverlaps(lines, true)
    val res2 = findNumOverlaps(lines, false)
    println("Answer 1: ${res1}\nAnswer 2: ${res2}")
}

File: rust/src/day05.rs
use crate::common;

extern crate num;
use num::signum;
use std::collections::HashMap;
use itertools::Itertools;

type Point = (i32, i32);
type Line = (Point, Point);

fn parse_pt(pt: &str) -> Point {
    pt.split(",")
        .map(|x| x.parse::<i32>().unwrap())
        .tuple_windows()
        .next()
        .unwrap()
}

fn parse_line(line: &str) -> Line {
    line.split(" -> ")
        .map(parse_pt)
        .tuple_windows()
        .next()
        .unwrap()
}

fn stroke(
    (x1, y1): &Point,
    (x2, y2): &Point,
    ptmap: &mut HashMap<Point, u32>,
    skip_diagonals: bool,
) {
    let dx = signum(x2 - x1);
    let dy = signum(y2 - y1);
    if skip_diagonals && dx != 0 && dy != 0 {
        return;
    }
    let mut cx = *x1;
    let mut cy = *y1;
    loop {
        *ptmap.entry((cx, cy)).or_insert(0) += 1;
        if cx == *x2 && cy == *y2 {
            break;
        }
        cx += dx;
        cy += dy;
    }
}

fn find_num_overlaps(lines: &Vec<Line>, skip_diagonals: bool) -> usize {
    let mut ptmap: HashMap<Point, u32> = HashMap::new();
    for (a, b) in lines {
        stroke(a, b, &mut ptmap, skip_diagonals);
    }
    ptmap.values().filter(|&x| *x > 1).count()
}

pub(crate) fn solution() {
    let lines: Vec<Line> = common::read_lines(&common::data_file(5))
        .unwrap()
        .iter()
        .map(|s| parse_line(&s))
        .collect();
    let res1 = find_num_overlaps(&lines, true);
    let res2 = find_num_overlaps(&lines, false);
    println!("Answer 1: {}\nAnswer 2: {}", res1, res2);
}
